packages/esm-appointments-app/src/appointments/common-components/end-appointment-modal.component.tsx

```diff
@@ -0,0 +1,104 @@
+import React from 'react';
+import { useTranslation } from 'react-i18next';
+import { useVisit, updateVisit, parseDate, showSnackbar } from '@openmrs/esm-framework';
+import { Button, ModalBody, ModalFooter, ModalHeader } from '@carbon/react';
+import { changeAppointmentStatus } from '../../patient-appointments/patient-appointments.resource';
+import { useMutateAppointments } from '../../form/appointments-form.resource';
+
+interface EndAppointmentModalProps {
+  patientUuid: string;
+  appointmentUuid: string;
+  closeModal: () => void;
+}
+
+const EndAppointmentModal: React.FC<EndAppointmentModalProps> = ({ patientUuid, appointmentUuid, closeModal }) => {
+  const { activeVisit, mutate } = useVisit(patientUuid);
+  const { t } = useTranslation();
+  const { mutateAppointments } = useMutateAppointments();
+
+  const endAppointment = () => {
```

Comment:
ideally, switching the appointment status and closing the visit would happen transactionally, but until we have backend functionality to do this, in case of an error, this should at least properly report back what worked/didn't worked.